fedb/api-server.md

```diff
@@ -0,0 +1,180 @@
+        "value1",
+        111,
+        1.4,
+        "2021-04-27",
+        1620471840256,
+        true
+    ]]
+
+}
+```
+response:
+```
+{
+    "code": 0,
+    "msg": "ok"
+}
+```
+
+|code|message|
+|--|--|
+|0|ok|
+|1|get insert information failed|
+
+### Execute Procedure 
+reqeust url: http://ip:port/dbs/{db_name}/procedures/{procedure_name}  
+http method: POST  
+request body: 
+```
+{
+    "common_cols":["value1", "value2"],
+    "input": [["value0", "value3"],["value0", "value3"]],
+    "need_schema": true
+}
+```
+response:
+```
+{
+    "code": 0,
+    "msg": "ok",
+    "data": {
+        "schema": [{"name": "field1", "type": "bool"}, {"name": "field2", "type": "int"}, {"name":"field3", "type": "bigint"}],
+        "data": [["value1", "value2"], [...]],
+        "common_cols_data": ["value3"]
+    }
+}
+```
+|code|message|
+|--|--|
+|0|ok|
+|-1|execute procedure failed|
+
+Note: If there is no common col, the reponse json will not contain common_cols_data field
+
+### Get Procedure
+request url: http://ip:port/dbs/{db_name}/procedures/{procedure_name}   
+http method: Get  
+response:
+```
+{
+    "code": 0,
+    "msg": "ok",
+    "data":{
+        "name": "procedure_name",
+        "procedure": "xxxxx",
+        "input_schema": [{"name":"field1", "type":"bool"}, {"name":"filed1", "type":"int"}],
+        "input_common_cols": ["field1", "field2"],
+        "output_schema": [{"name":"field1", "type":"bool"}, {"name":"field2", "type":"int32"}, {"name":"field3", "type":"bigint"}],
+        "output_common_cols":["field3"]
+        "tables": ["table1", "table2"]
+    }
+}
+```
+|code|message|
+|--|--|
+|0|ok|
+|-1|procedure not found|
+
+Note: If there is no common col, the reponse json will not contain input_common_cols/output_common_cols field
+
+## Server design
+
+We use [brpc HTTP Service](https://github.com/apache/incubator-brpc/blob/master/docs/en/http_service.md) framework to tackle the http request. So what need to do is forwarding the reques to FEDB server with fedb sdk in brpc http service implementation.
+
+1. Design proto file.  
+    ```
+    option cc_generic_services = true;
+
+    message HttpRequest { };
+    message HttpResponse { };
+
+    service APIService {
+        rpc ProcessSQL(HttpRequest) returns (HttpResponse);
+    }
+    ```
+2. Implement the service  
+    ```
+    class APIServiceImpl : public APIService {
+      public:
+        bool Init() {
+            // init the sdk
+            ...
+        }
+        virtual void Process(google::protobuf::RpcController* controller,
+                                const HttpRequest* request,
+                                HttpResponse* response,
+                                google::protobuf::Closure* done) {
+            // parse url and dispatch
+        }
+    };
+    ```
+3. Add mapping
+    
+    ```
+    if (server.AddService(&api_svc,
+                      brpc::SERVER_DOESNT_OWN_SERVICE,
+                      "/dbs/*  => Process" != 0) {
```

Comment:
The impl is "/* => Process", which one needs fix?

Reply:
done